### PR TITLE
feat(core,ui,vite-plugin): add option to group output by scenes

### PR DESCRIPTION
**Pull request by misclicl**
This update adds an option to the Video Settings tab that allows users to organise their project output by scenes. The options is disabled by default.

### CODE REVIEW

**Comment by misclicl** [comment]
damn, sorry for the mess, but I can't figure how I'm supposed to link the issue (https://github.com/motion-canvas/motion-canvas/issues/462) to this. i haven't used Github in a while. does the name have to be identical?

File: packages/ui/src/components/settings/RenderingSettings.tsx [hunk 1]
```diff
@@ -69,6 +69,16 @@ export function RenderingSettings() {
           />
         </Group>
       )}
+      <Group>
+        <Label>{meta.rendering.groupByScene.name}</Label>
```
Comment (misclicl) [inline packages/ui/src/components/settings/RenderingSettings.tsx:73]: i'm not sure if this meant to be pulled from the `meta` to be honest. `MetaField` documentation mentions something like this: https://github.com/motion-canvas/motion-canvas/blob/1c259d0d574bb56dbc8bc448300d9b94ee4d0bc4/packages/core/src/meta/MetaField.ts#L25-L31
while no other fields use it. depending on your answer i can either change it all to read from `MetaField` value, or use string literal here instead.

**Comment by misclicl** [comment]
hmm, now thinking about it i realised that i probably want to reset the frame number in between scenes.

File: packages/ui/src/components/settings/RenderingSettings.tsx [hunk 2]
```diff
@@ -69,6 +69,16 @@ export function RenderingSettings() {
...
+        <Input
+          type="checkbox"
+          checked={Boolean(state.groupByScene)}
```
Comment (aarthificial) [inline packages/ui/src/components/settings/RenderingSettings.tsx:76]: The cast is not necessary:
```suggestion
          checked={state.groupByScene}
```
The type system prevents anything other than a boolean to be set as the value of this field.
If you want to make sure that the value read from the meta file is correct, you should create a new `BoolMetaField` class and override the `parse` method.

File: packages/core/src/vite-env.d.ts [hunk 3]
```diff
@@ -7,11 +7,13 @@ declare module 'vite/types/customEvent' {
...
       data: string;
-      mimeType: string;
-      project: string;
-      isStill: boolean;
+      meta: {
+        name: string;
+        subDirectories: string[];
+        mimeType: string;
+        frameNumber: number;
+      };
```
Comment (aarthificial) [inline packages/core/src/vite-env.d.ts:16]: I'm not sure what's the point of wrapping these things in an additional object.

**Review by aarthificial — CHANGES_REQUESTED** [review]

File: packages/ui/src/components/settings/RenderingSettings.tsx (hunk 2 above)
Comment (misclicl) [reply packages/ui/src/components/settings/RenderingSettings.tsx:76]: you're right, not sure why i added `Boolean` here. will remove. although idk about your suggestion with `BoolMetaField`. does it make sense to create it now? I initialise `groupByScene` with a boolean value in `createProjectMetadata`. Unless someone decides to manually modify meta files this seems useless.

File: packages/core/src/app/FrameSaver.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
+export class FrameSaver {
```
[the remark this answers is not in the document]
Comment (misclicl) [reply packages/core/src/app/FrameSaver.ts]: I definitely overcomplicated lot of things here. thanks for the review. `FrameSaver` class is removed.

File: packages/ui/src/components/settings/RenderingSettings.tsx (hunk 1 above)
Comment (misclicl) [reply packages/ui/src/components/settings/RenderingSettings.tsx]: alright, I'll just use string literal here then.

File: packages/core/src/app/ImageExporter.ts [hunk 5]
```diff
@@ -56,13 +60,19 @@ export class ImageExporter implements Exporter {
...
+        {
+          frame,
+          fileType: this.fileType,
+          projectName: this.projectName,
+          groupByScene: this.groupByScene,
+          sceneName,
+        },
+      );
+
+      return Promise.resolve();
```
[the remark this answers is not in the document]
Comment (misclicl) [reply packages/core/src/app/ImageExporter.ts]: removed.

File: packages/core/src/vite-env.d.ts (hunk 3 above)
Comment (misclicl) [reply packages/core/src/vite-env.d.ts]: changed it to a flat structure now.

**Comment by misclicl** [comment]
@aarthificial I addressed everything.

**Comment by aarthificial** [comment]
Thanks!
To address your comments that I missed before:
> does the name have to be identical?

If the pr closes an issue, the commit should contain this in its body:
```
Closes: #ISSUE_ID
```
But I can add this when merging so no worries.

> hmm, now thinking about it i realised that i probably want to reset the frame number in between scenes.

I think it's a good idea too. You can calculate the frame number relative to the scene in the `Renderer`:
```ts
const frame = this.playback.frame - this.playback.currentScene.firstFrame;
```

**Comment by nick-1666** [comment]
Thanks for working on this. I wanted this for the longest time :)

**Comment by misclicl** [comment]
> I think it's a good idea too. You can calculate the frame number relative to the scene in the Renderer:
> ```
> const frame = this.playback.frame - this.playback.currentScene.firstFrame;
> ```
i tried this, but `this.playback.currentScene.firstFrame` seems to return `0` regardless of the current scene. i can look into it a bit later.

**Comment by aarthificial** [comment]
image
image

**Comment by misclicl** [comment]
@aarthificial, the issue with `0` is caused by the `PlaybackManager` instance used in `Renderer` never updating the `firstFrame` variable because the `recalculate` function is never called. I tried calling `recalculate` during the `render`, but it caused the rendering to become stuck at the very end. While I can investigate further, do you have any tips for me?

**Comment by aarthificial** [comment]
Rendering will always get stuck when using `ui:serve` because there's no backend in place to handle the export.
`playback.recalculate()` should be called in the `run` method, after `reloadScenes`.

**Comment by misclicl** [comment]
I was running all core, UI, and plugins in watch mode with template:serve for this issue, so that wasn't the problem. The real reason was that I messed up ImageRenderer 🤦 Anyway, everything is working properly now.

**Review by aarthificial — CHANGES_REQUESTED** [review]
Requesting changes because there are merge conflicts

**Review by aarthificial — APPROVED** [review]

**Comment by aarthificial** [comment]
Awesome, thanks for the PR!

**Pull request merged**